File: mindspore/_extends/graph_kernel/expanders/tile.py
# Copyright 2020-2021 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ===========================================================================
"""generate json desc for Tile"""
from mindspore._extends.graph_kernel.model.model import GraphKernelUnsupportedException as GKException
from mindspore._extends.graph_kernel.model.model import DataFormat as DF
from ._utils import Expander, ExpanderInfoValidator as VLD


@VLD.add_format(DF.DEFAULT)
@VLD.check_attrs('multiples')
class Tile(Expander):
    """Tile expander"""

    def _get_output_shape(self):
        """Get output shape"""
        shape = self.inputs[0].shape
        multiples = self.attrs["multiples"]

        shape = list(shape)
        multiples = list(multiples)
        diff_len = len(multiples) - len(shape)
        if diff_len < 0:
            raise GKException("Dimensions of multiples{} < dimensions of input{} in Tile".format(multiples, shape))
        if diff_len > 0:
            for _ in range(diff_len):
                shape.insert(0, 1)

        output_shape = []

        for sh, mul in list(zip(shape, multiples)):
            if sh != 1 and mul != 1:
                raise GKException("Tile op in expander only Support Automatic Broadcast!")
            dim = sh * mul
            output_shape.append(dim)
        return output_shape

    def _expand(self, graph_builder):
        input_x = self.inputs[0]
        output_shape = self._get_output_shape()

        result = graph_builder.emit('BroadcastTo', [input_x], attrs={'shape': output_shape})
        return result

File: mindspore/ccsrc/backend/common/graph_kernel/expanders/tile.cc
/**
 * Copyright 2023 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <memory>
#include <vector>

#include "mindapi/base/shape_vector.h"
#include "backend/common/graph_kernel/expanders/op_desc_registry.h"
#include "ops/ops_func_impl/tile.h"

namespace mindspore::graphkernel::expanders {
class Tile : public OpDesc {
 public:
  Tile() {
    std::initializer_list<std::string> attrs{"dims"};
    (void)validators_.emplace_back(std::make_unique<CheckAttr>(attrs));
  }
  ~Tile() = default;

 protected:
  bool CheckInputs() override {
    const auto &x = inputs_info_[0];
    auto x_shape = x.shape;
    auto dims = GetAxisList(attrs_["dims"]);
    if (IsDynamicRank(x_shape)) {
      MS_LOG(INFO) << "Skip dynamic rank case";
      return false;
    }
    ops::AdaptShapeAndMultipies(&x_shape, &dims);
    for (size_t i = 0; i < x_shape.size(); ++i) {
      if (x_shape[i] != 1 && dims[i] != 1) {
        MS_LOG(INFO) << "For 'Tile', input.shape = " << x.shape << ", dims = " << GetAxisList(attrs_["dims"])
                     << ", both value are not equal to 1 in matched-index " << i
                     << ", which can not be replaced by 'BroadcastTo'";
        return false;
      }
    }
    return true;
  }

  NodePtrList Expand(const NodePtrList &inputs) override {
    const auto &x = inputs[0];
    auto x_shape = x->shape;
    auto dims = GetAxisList(attrs_["dims"]);
    ops::AdaptShapeAndMultipies(&x_shape, &dims);
    ShapeVector out_shape = std::move(dims);
    for (size_t i = 0; i < x_shape.size(); ++i) {
      out_shape[i] *= x_shape[i];
    }

    auto result = gb.BroadcastTo(x, out_shape);
    return {result};
  }
};
EXPANDER_OP_DESC_REGISTER("Tile", Tile);
}  // namespace mindspore::graphkernel::expanders
